packages/grid/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -116,13 +116,35 @@ export const useGridVirtualScroller = (props: UseGridVirtualScrollerProps) => {
 
     const { top, left } = scrollPosition.current!;
 
-    const firstRowIndex = getIndexFromScroll(top, rowsMeta.positions);
+    const firstRowIndex = Math.min(
+      apiRef.current.unstable_getLastMeasuredRowIndex(),
+      getIndexFromScroll(top, rowsMeta.positions),
+    );
     const lastRowIndex = rootProps.autoHeight
       ? firstRowIndex + currentPage.rows.length
       : getIndexFromScroll(top + rootRef.current!.clientHeight!, rowsMeta.positions);
 
-    const firstColumnIndex = getIndexFromScroll(left, columnPositions);
-    const lastColumnIndex = getIndexFromScroll(left + containerWidth!, columnPositions);
+    let hasRowWithAutoHeight = false;
+    let firstColumnIndex = 0;
+    let lastColumnIndex = columnPositions.length;
+
+    const [firstRowToRender, lastRowToRender] = getRenderableIndexes({
+      firstIndex: firstRowIndex,
+      lastIndex: lastRowIndex,
+      minFirstIndex: 0,
+      maxLastIndex: currentPage.rows.length,
+      buffer: rootProps.rowBuffer,
+    });
+
+    for (let i = firstRowToRender; i < lastRowToRender && !hasRowWithAutoHeight; i += 1) {
+      const row = currentPage.rows[i];
+      hasRowWithAutoHeight = apiRef.current.unstable_rowHasAutoHeight(row.id);
```

Comment:
Why not 

```js
hasRowWithAutoHeight = rows.slice(firstRowToRender, lastRowToRender).some(row => apiRef.current.unstable_rowHasAutoHeight(row.id))
```

I get a bit confused because I missed the `!hasRowWithAutoHeight`

Reply:
The for loop is a small optimization to keep complexity at O(n). Using `Array.slice` with `Array.some` has complexity O(n²) because it first iterates between the two indices to create the sliced array, then iterates again over it to test if it satistifies the condition.